qlib/data/cache.py

```diff
@@ -39,33 +40,123 @@ class QlibCacheException(RuntimeError):
     pass
 
 
-class MemCacheUnit(OrderedDict):
+class MemCacheUnit(abc.ABC):
     """Memory Cache Unit."""
 
-    # TODO: use min_heap to replace ordereddict for better performance
-
     def __init__(self, *args, **kwargs):
-        self.size_limit = kwargs.pop("size_limit", None)
-        # limit_type: check size_limit type, length(call fun: len) or size(call fun: sys.getsizeof)
+        self.size_limit = kwargs.pop("size_limit", 0)
         self.limit_type = kwargs.pop("limit_type", "length")
-        super(MemCacheUnit, self).__init__(*args, **kwargs)
-        self._check_size_limit()
+
+        assert self.limit_type in ["length", "sizeof"], ValueError("limit_type shoule be one of ['length', 'sizeof']")
+        assert self.size_limit >= 0, ValueError("size_limit shoule not be negative.The default 0 means unlimited!")
+
+        # limit_flag: whether to popitem or not
+        self._limit_flag = 1 if self.size_limit > 0 else 0
+        self._size = 0
+        self.od = OrderedDict()
 
     def __setitem__(self, key, value):
-        super(MemCacheUnit, self).__setitem__(key, value)
-        self._check_size_limit()
+        # TODO: thread safe?__setitem__ failure might cause inconsistent size?
+
+        # precalculate the size after od.__setitem__
+        self._adjust_size(key, value)
+
+        self.od.__setitem__(key, value)
+
+        # move the key to end,make it latest
+        self.od.move_to_end(key)
+
+        # pop the oldest items beyond size limit
+        while self._size * self._limit_flag > self.size_limit:
+            self.popitem(last=False)
 
     def __getitem__(self, key):
-        value = super(MemCacheUnit, self).__getitem__(key)
-        super(MemCacheUnit, self).__delitem__(key)
-        super(MemCacheUnit, self).__setitem__(key, value)
-        return value
+        v = self.od.__getitem__(key)
+        self.od.move_to_end(key)
+        return v
+
+    def __contains__(self, key):
+        return key in self.od
+
+    def __len__(self):
+        return self.od.__len__()
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}<limit_type:{self.limit_type} size_limit:{self.size_limit if self.is_limited else 'no limit'} total_size:{self._size}>\n{self.od.__repr__()}"
+
+    def set_limit_size(self, limit):
+        self.size_limit = limit
+
+    def set_limit(self, is_limited=True):
+        self._limit_flag = 1 if is_limited else 0
+
+    @property
+    def is_limited(self):
+        return bool(self._limit_flag)
+
+    @property
+    def total_size(self):
+        return self._size
+
+    def clear(self):
+        self._size = 0
+        self.od.clear()
+
+    @abc.abstractmethod
+    def _adjust_size(self, key, value):
```

Comment:
I think all the differences between the subclasses are the way to calculate the size of value.
Other parts  can be shared.
I think we could make  only one method to calculate size of value overridable. Other logics can be shared in the base class (e.g.  pop, popitem, adjust_size ... ).

Reply:
got it